--- ./src/.minecraft_orig/net/minecraft/src/Gui.java	2013-05-22 01:14:06.483046000 -0500
+++ ./modsrc/minecraft/net/minecraft/src/Gui.java	2013-05-22 02:42:15.147096917 -0500
@@ -5,6 +5,7 @@
 public class Gui
 {
     protected float zLevel = 0.0F;
+    public boolean _isForOculus = false;
 
     protected void drawHorizontalLine(int par1, int par2, int par3, int par4)
     {
@@ -119,22 +120,40 @@
         par1FontRenderer.drawStringWithShadow(par2Str, par3, par4, par5);
     }
 
-    /**
-     * Draws a textured rectangle at the stored z-value. Args: x, y, u, v, width, height
-     */
-    public void drawTexturedModalRect(int par1, int par2, int par3, int par4, int par5, int par6)
+    public void drawTexturedModalRect(double x, double y, int u, int v, int width, int height, float scaleFactor)
     {
         float var7 = 0.00390625F;
         float var8 = 0.00390625F;
         Tessellator var9 = Tessellator.instance;
         var9.startDrawingQuads();
-        var9.addVertexWithUV((double)(par1 + 0), (double)(par2 + par6), (double)this.zLevel, (double)((float)(par3 + 0) * var7), (double)((float)(par4 + par6) * var8));
-        var9.addVertexWithUV((double)(par1 + par5), (double)(par2 + par6), (double)this.zLevel, (double)((float)(par3 + par5) * var7), (double)((float)(par4 + par6) * var8));
-        var9.addVertexWithUV((double)(par1 + par5), (double)(par2 + 0), (double)this.zLevel, (double)((float)(par3 + par5) * var7), (double)((float)(par4 + 0) * var8));
-        var9.addVertexWithUV((double)(par1 + 0), (double)(par2 + 0), (double)this.zLevel, (double)((float)(par3 + 0) * var7), (double)((float)(par4 + 0) * var8));
+        var9.addVertexWithUV((double)(x + 0), (double)(y + height * scaleFactor), (double)this.zLevel, (double)((float)(u + 0) * var7), (double)((float)(v + height) * var8));
+        var9.addVertexWithUV((double)(x + width * scaleFactor), (double)(y + height * scaleFactor), (double)this.zLevel, (double)((float)(u + width) * var7), (double)((float)(v + height) * var8));
+        var9.addVertexWithUV((double)(x + width * scaleFactor), (double)(y + 0), (double)this.zLevel, (double)((float)(u + width) * var7), (double)((float)(v + 0) * var8));
+        var9.addVertexWithUV((double)(x + 0), (double)(y + 0), (double)this.zLevel, (double)((float)(u + 0) * var7), (double)((float)(v + 0) * var8));
         var9.draw();
     }
 
+    public void drawTexturedModalRect(int x, int y, int u, int v, int width, int height, float scaleFactor)
+    {
+        float var7 = 0.00390625F;
+        float var8 = 0.00390625F;
+        Tessellator var9 = Tessellator.instance;
+        var9.startDrawingQuads();
+        var9.addVertexWithUV((double)(x + 0), (double)(y + height * scaleFactor), (double)this.zLevel, (double)((float)(u + 0) * var7), (double)((float)(v + height) * var8));
+        var9.addVertexWithUV((double)(x + width * scaleFactor), (double)(y + height * scaleFactor), (double)this.zLevel, (double)((float)(u + width) * var7), (double)((float)(v + height) * var8));
+        var9.addVertexWithUV((double)(x + width * scaleFactor), (double)(y + 0), (double)this.zLevel, (double)((float)(u + width) * var7), (double)((float)(v + 0) * var8));
+        var9.addVertexWithUV((double)(x + 0), (double)(y + 0), (double)this.zLevel, (double)((float)(u + 0) * var7), (double)((float)(v + 0) * var8));
+        var9.draw();
+    }
+
+    /**
+     * Draws a textured rectangle at the stored z-value. Args: x, y, u, v, width, height
+     */
+    public void drawTexturedModalRect(int x, int y, int u, int v, int width, int height)
+    {
+        drawTexturedModalRect(x, y, u, v, width, height, 1.0f);
+    }
+
     public void drawTexturedModelRectFromIcon(int par1, int par2, Icon par3Icon, int par4, int par5)
     {
         Tessellator var6 = Tessellator.instance;
