--- a/net/minecraft/src/GuiContainerCreative.java
+++ b/net/minecraft/src/GuiContainerCreative.java
@@ -548,9 +548,9 @@
     /**
      * Handles mouse input.
      */
-    public void handleMouseInput()
+    public void handleMouseInput(int displayWidth, int displayHeight, boolean splitView)
     {
-        super.handleMouseInput();
+        super.handleMouseInput(displayWidth, displayHeight, splitView);
         int var1 = Mouse.getEventDWheel();
 
         if (var1 != 0 && this.needsScrollBars())
