--- a/net/minecraft/src/GuiChat.java
+++ b/net/minecraft/src/GuiChat.java
@@ -132,9 +132,9 @@
     /**
      * Handles mouse input.
      */
-    public void handleMouseInput()
+    public void handleMouseInput(int displayWidth, int displayHeight, boolean splitView)
     {
-        super.handleMouseInput();
+        super.handleMouseInput(displayWidth, displayHeight, splitView);
         int var1 = Mouse.getEventDWheel();
 
         if (var1 != 0)
