--- ./src/.minecraft_orig/net/minecraft/src/Profiler.java	2013-05-22 01:14:06.631046000 -0500
+++ ./modsrc/minecraft/net/minecraft/src/Profiler.java	2013-05-22 02:42:15.147096917 -0500
@@ -56,7 +56,7 @@
      */
     public void startSection(String par1Str)
     {
-        if (Config.getGameSettings().showDebugInfo)
+        if (Config.getGameSettings().showDebugInfo > 0)
         {
             if (this.startTickNano == 0L && par1Str.equals("tick"))
             {
